hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/keyvalue/KvMetadataAvroResolver.java

```diff
@@ -99,24 +102,27 @@ public Stream<MappingField> resolveAndValidateFields(
             Map<String, String> options,
             InternalSerializationService serializationService
     ) {
-        if (userFields.isEmpty()) {
-            throw QueryException.error("Column list is required for Avro format");
-        }
         Map<QueryPath, MappingField> fieldsByPath = extractFields(userFields, isKey);
+        for (QueryPath path : fieldsByPath.keySet()) {
+            if (path.isTopLevel()) {
+                throw QueryException.error("Cannot use the '" + path + "' field with Avro serialization");
+            }
+        }
 
         Schema schema = getSchema(fieldsByPath, options, isKey);
         if (schema != null && options.containsKey("schema.registry.url")) {
             throw new IllegalArgumentException("Inline schema cannot be used with schema registry");
         }
-        for (QueryPath path : fieldsByPath.keySet()) {
-            if (path.isTopLevel()) {
-                throw QueryException.error("Cannot use the '" + path + "' field with Avro serialization");
+
+        if (fieldsByPath.isEmpty()) {
```

Comment:
if in MAPPING I explicitly reference only fields from `key` and provide schema for both `key` and `value`, then fields from `value` are added.

```java
        kafkaMapping("aaa", ID_SCHEMA, ID_SCHEMA)
                .fields("\"idkey\" INTEGER EXTERNAL NAME \"__key.id\"")
                .create();
        sqlService.execute("select get_ddl('relation', 'aaa')").stream().forEach(System.out::println);
```
produces:
```sql
CREATE OR REPLACE EXTERNAL MAPPING "hazelcast"."public"."aaa" EXTERNAL NAME "aaa" (
  "idkey" INTEGER EXTERNAL NAME "__key.id",
  "id" INTEGER EXTERNAL NAME "this.id"
)
TYPE "Kafka"
OBJECT TYPE "Topic"
OPTIONS (
  'keyFormat'='avro',
  'valueFormat'='avro',
  'bootstrap.servers'='PLAINTEXT://localhost:49238',
  'auto.offset.reset'='earliest',
  'keyAvroSchema'='{"type":"record","name":"sql","namespace":"jet","fields":[{"name":"id","type":["null","int"],"default":null}]}',
  'valueAvroSchema'='{"type":"record","name":"sql","namespace":"jet","fields":[{"name":"id","type":["null","int"],"default":null}]}'
)
```

Is it common behavior also for other serialization formats?

Reply:
Key and value field resolution are independent, so it is expected and common for all formats. Currently, in `KvMetadataResolver`s, the list of all mapping fields is checked against emptiness to decide field resolution, instead of only key or value fields. This will be resolved in HZ-3561.

Reply:
>  it is expected and common for all formats

I see that it is not true for` KvMetadataJavaResolver.resolveAndValidateFields` and `MetadataPortableResolver`: they checks for emptiness of `userFields` not `fieldsByPath`. Some other resolvers (including Compact) require explicit fields. So this needs to be fixed only for Avro, possibly as simple as:
```suggestion
        if (userFields.isEmpty()) {
```

Reply:
Having no fields for key/value doesn't make sense. There are two ways of solving it:
1. Throw an exception saying "mapping has incomplete fields."
2. Do the field resolution for the part(s) it is necessary.

I have chosen the second option because I see no reason for the "specify everything or nothing" behavior when we have the ability to resolve key/value fields independently. The other formats clearly don't implement the first option, and that's why I assumed that they incorrectly implement the second option. So, from my point of view, the ones that need to be fixed are Java and Portable formats, and the statement "it is expected and common for all formats" is true. Finally, please keep in mind that mappings can have mixed serialization formats for key and value, which is just another reason I don't like the "specify everything or nothing" behavior.

Reply:
> I see no reason for the "specify everything or nothing" behavior when we have the ability to resolve key/value fields independently
I think the correct behavior is: when you specify list of columns they are all and only columns that you want. Magically adding some columns is unexpected and hard to understand (why some columns were added and others not). And this would work only for key-value connectors which adds yet another surprise.
If we wanted the behavior you describe, it should be requested explicitly, eg. `CREATE MAPPING WITH DEFAULT [KEY|VALUE] FIELDS`.

If the column list is invalid (eg. if mapping without value columns would be invalid or is logically impossible) this should be checked explicitly. But, at least for reading mapping with only key fields or only value fields should be fine, so I do not see any reason to forbid that.

> keep in mind that mappings can have mixed serialization formats for key and value

For Java + Compact you could create mapping with explicit fields referencing only Java fields. That would create empty schema for compact part. Maybe that is fine (is empty schema valid?) or if not, it should be checked: for compact we might require non-empty `fieldsByPath`.

For Avro, we could forbid situation when `userFields` is non empty but schema is not given and `fieldsByPath` is empty. That would produce empty `jet.sql` schema but that might be fine or not.

Reply:
[`a308af0c`..`721a8f7a`][1]

[1]: https://github.com/hazelcast/hazelcast/compare/a308af0cc38ce11267df749833964cec397ccce6..721a8f7a755ea67ac69c253dc3667a5df64d57c5